contrib/repro_test.bzl

```diff
@@ -0,0 +1,222 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test rule for verifying a given container_image target produces a
+   deterministic Docker image.
+
+This rule builds the given container_image twice in isolation to determine if
+the given image is deterministic.
+It does so by mounting the provided Bazel project files into the 'base' image.
+Next it starts two containers and builds the supplied container_image target
+(being 'image') inside each of the containers. The rule then extracts the
+output files for the test image. Finally, both extracted images are compared.
+A test image is considered to be reproducible when both instances of the same
+image have the same digest and ID. On a mismatch of one of those values, the
+test fails and the container_diff tool
+(https://github.com/GoogleContainerTools/container-diff#container-diff)
+produces the summary of their differences.
+
+Args:
+    image: A container_image target relative to the project's root to test
+           for determinism.
+    base: An image target to build and reproduce the 'image 'in.
+          This image must have Bazel and docker installed and available in the PATH.
+    workspace_file: The WORKSPACE file of the project containing the 'image'
+                    target to help detect project's root path.
+    container_diff_args: (optional) Args to the container_diff tool as specified here:
+                         https://github.com/GoogleContainerTools/container-diff#quickstart
+
+Example:
+
+container_repro_test(
+    name = "set_cmd_repro_test",
+    image = "//tests/container:set_cmd",
+    base = "@bazel_latest//image",
+    workspace_file = "//:WORKSPACE",
+    container_diff_args: ["history", "file", "size", "rpm", "apt", "pip", "node"],
+)
+"""
+
+load("@base_images_docker//util:run.bzl", _extract = "extract")
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//container:container.bzl", _container = "container")
+load("//skylib:filetype.bzl", container_filetype = "container")
+
+def _impl(ctx):
+    """Core implementation of container_repro_test"""
+
+    name = ctx.attr.name
+    workspace_file = ctx.file.workspace_file
+    proj_root = ctx.actions.declare_file(name + "_src_project_root.txt")
+
+    # Record the absolute path to the provided WORKSPACE file and assume it to
+    # be the root of the Bazel project to build the 'image' in.
+    ctx.actions.run_shell(
+        command = "readlink -f %s | xargs dirname >> %s" % (workspace_file.path, proj_root.path),
+        inputs = [workspace_file],
+        outputs = [proj_root],
+    )
+
+    # Tar the current source project files to mount into the image that will
+    # be used to build and repro the 'image' in.
+    proj_tar = ctx.actions.declare_file(name + "_src_project.tar")
+    ctx.actions.run_shell(
+        command = "tar -cvf %s \"$(cat %s)\"" % (proj_tar.path, proj_root.path),
+        inputs = [proj_root],
```

Comment:
I really dont think you should need a file to just store the path of the dir for the WORKSPACE file if this is the only place where you are using it. When I talked before about using a file it was because I was thinking of one repo rule to identify the root and then a different rule to read that file.

Reply:
As discussed offline, this seems the only way to do it. I'm able to get the absolute path to the WORKSPACE file only via the `ctx.actions.run_shell`, storing the path in a declared file. As I'm unable to read the file to store the path in a variable, I have to refer to the file content when the path needs to be used.

Reply:
sgtm